File: go-RayTracer/hittable.go
package main

// Material is an interface that represents a material type
type Material interface {
	Scatter(rIn Ray, rec HitRecord) (bool, Color, Ray)
}

// HitRecord stores information about a ray-object intersection
type HitRecord struct {
	P         Point3
	Normal    Vec3
	Mat       Material
	T         float64
	FrontFace bool
}

// SetFaceNormal sets the hit record normal vector based on the ray direction and the outward normal
func (rec *HitRecord) SetFaceNormal(r Ray, outwardNormal Vec3) {
	// Sets the hit record normal vector.
	// NOTE: the parameter `outwardNormal` is assumed to have unit length.

	rec.FrontFace = Dot(r.Direction(), outwardNormal) < 0
	if rec.FrontFace {
		rec.Normal = outwardNormal
	} else {
		rec.Normal = outwardNormal.Neg()
	}
}

// Hittable is an interface for objects that can be hit by a ray
type Hittable interface {
	Hit(r Ray, rayT Interval, rec *HitRecord) bool
}

File: ts-RayTracer/src/hittable.ts
import { Point3, Vec3 } from './vec3';
import { Ray } from './ray';
import { Interval } from './interval';

export interface Material {
  scatter(rIn: Ray, rec: HitRecord): { attenuation: Vec3; scattered: Ray } | null;
}

export class HitRecord {
  p!: Point3;
  normal!: Vec3;
  mat!: Material;
  t!: number;
  frontFace!: boolean;

  setFaceNormal(r: Ray, outwardNormal: Vec3): void {
    this.frontFace = dotLessThanZero(r, outwardNormal);
    this.normal = this.frontFace ? outwardNormal : outwardNormal.neg();
  }
}

function dotLessThanZero(r: Ray, outwardNormal: Vec3): boolean {
  const d = r.direction();
  return d.x() * outwardNormal.x() + d.y() * outwardNormal.y() + d.z() * outwardNormal.z() < 0;
}

export interface Hittable {
  hit(r: Ray, rayT: Interval, rec: HitRecord): boolean;
}


